vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,200 @@
+- `integer` is written as is to the data. The class is stored in the meta data.
+- `numeric` is written as is to the data. The class is stored in the meta data.
+- `complex` is written as is to the data. The class is stored in the meta data.
+- `character` is written as is and unquoted to the data. The class is stored in the meta data.
+- `factor` is  stored as its indices in the data. The labels of levels are stored in the meta data.
+- `POSIXct` is written as an integer to the data. The class and the origin are stored in the meta data.
+
+```{r}
+# Create dummy data
+x <- data.frame(
+    x = LETTERS,
+    y = factor(
+        sample(c("a", "b", NA), 26, replace = TRUE),
+        levels = c("a", "b", "c")
+    ),
+    z = c(NA, 1:25),
+    abc = c(rnorm(25), NA),
+    xyz = complex(real = rnorm(26), imaginary = rnorm(26)),
+    def = sample(c(TRUE, FALSE, NA), 26, replace = TRUE),
+    timestamp = seq(
+        as.POSIXct("2018-01-01"),
+        as.POSIXct("2019-01-01"),
+        length = 26
+    ),
+    stringsAsFactors = FALSE
+)
+str(x)
+```
+
+Another important part of the preprocessing is ordering both the variables as the observations. This is determined when the file is created and stored in the meta data. When the file is overwritten, the ordering is based on the existing meta data. Therefore it is important to think carefully about the ordering and make sure that is how you want it to stay **before** you create your first commit. The ordering of the observations is specified through the `sorting` argument. It holds a vector of variable names which are used in the ordering. Make sure that you add enough sorting variables to avoid ties. `sorting` also defines the order of the variables, the order of the other variables is taken from the `data.frame`.
+
+The data file (".tsv") and the meta data file (".yml") are automatically staged after writing. `write_delim_git()` returns the hashes of both the data file and the meta data file. The names of the vector contain the path of the files, relative to the root of the _git_ repository.
+
+```{r}
+status(data_repo)
+write_delim_git(x = x, file = "my_data", repo = data_repo, sorting = c("y", "x"))
+status(data_repo)
+status(repo)
+```
+
+Overwriting data which has different meta data than the original data will throw an error because this can potentially lead to large diffs. Differences in meta data occur when:
+
+- the sorting changes
+- variables are added, removed or renamed
+- the class of a variable changes
+- the levels of a factor changes
+
+However you can bypass this by using `override = TRUE`. This will completely ignore the existing meta data and create new meta data based on the new `data.frame`.
+
+```{r error=TRUE}
+y <- x
+y$extra <- x$x
+write_delim_git(y, file = "my_data", repo = data_repo, sorting = c("y", "x"))
+write_delim_git(
+    y, file = "my_data", repo = data_repo, sorting = c("y", "x"), 
+    override = TRUE
+)
+```
```

Comment:
I like the concept of the meta-data-versioning while having the option to override it, cool!